src/pose-sensor.js

```diff
@@ -0,0 +1,170 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import FusionPoseSensor from './sensor-fusion/fusion-pose-sensor.js';
+import { Vector3, Quaternion } from './math-util.js';
+
+// Frequency which the Sensors will attempt to fire their
+// `reading` functions at. Use 16.6ms to achieve 60FPS, since
+// we generally can't get higher without native VR hardware
+const SENSOR_FREQUENCY = 16.6;
+
+const X_AXIS = new Vector3(1, 0, 0);
+const Z_AXIS = new Vector3(0, 0, 1);
+
+let orientation = {};
+if (screen.orientation) {
+  orientation = screen.orientation;
+} else if (screen.msOrientation) {
+  orientation = screen.msOrientation;
+} else {
+  Object.defineProperty(orientation, "angle", {
+    get: () => { return (window.orientation || 0) }
+  });
+}
+
+// Quaternion to rotate from sensor coordinates to WebVR coordinates
+const SENSOR_TO_VR = new Quaternion();
+SENSOR_TO_VR.setFromAxisAngle(X_AXIS, -Math.PI / 2);
+SENSOR_TO_VR.multiply(new Quaternion().setFromAxisAngle(Z_AXIS, Math.PI / 2));
+
+/**
+ * An abstraction class around either using the new RelativeOrientationSensor,
+ * or `devicemotion` events with complimentary filter via fusion-pose-sensor.js.
+ */
+export default class PoseSensor {
+  constructor(config) {
+    this.config = config;
+    this.sensor = null;
+    this.fusionSensor = null;
+    this._out = new Float32Array(4);
+
+    // Can be 'sensor' (using RelativeOrientationSensor) or
+    // 'devicemotion' (using devicemotion events via FusionPoseSensor)
+    this.api = 'sensor';
+
+    // Store any errors from Sensors
+    this.error = null;
+
+    // Quaternions for caching transforms
+    this._sensorQ = new Quaternion();
+    this._worldToScreenQ = new Quaternion();
+    this._outQ = new Quaternion();
+
+    this._onSensorRead = this._onSensorRead.bind(this);
+    this._onSensorError = this._onSensorError.bind(this);
+    this._onOrientationChange = this._onOrientationChange.bind(this);
+
+    this._onOrientationChange();
+    this.init();
+  }
+
+  init() {
+    // Attempt to use the RelativeOrientationSensor from Generic Sensor APIs.
+    // First available in Chrome M63, this can fail for several reasons, and attempt
+    // to fallback to devicemotion. Failure scenarios include:
+    //
+    // * Generic Sensor APIs do not exist, fallback to devicemotion
+    // * Underlying sensor does not exist, no fallback possible.
+    // * Feature policy failure (in an iframe); no fallback.
+    // * Permission to sensor data denied; respect user agent, no fallback to devicemotion.
+    //   Browsers are heading towards disabling devicemotion when sensors are denied as well.
```

Comment:
I know it's getting verbose (and fantastic work + patience, btw), but could we add relevant URLs?

Reply:
Added the [Chrome deprecating on-by-default permissions](https://www.chromestatus.com/feature/5023919287304192) and [Feature Policy in WebXR for iframes](https://github.com/immersive-web/webxr/issues/86) as more info, good call! As a follow up, I'll add more information to the README (linked from webvr-polyfill as well) as there's a bit more to handle in the iframe scenario on the developers end, and to indicate what underlying APIs are being used now